### PR TITLE
[chip dv] Enable parallel meson invocation

### PR DESCRIPTION
**Pull request by sriyerg**
This update enables parallel meson invocations from simulations running
in parallel, effectively fixing #1708.

It uses the linux build-in `flock` utility to lock the common sw build directory during parallel meson invocations.

Signed-off-by: Srikrishna Iyer <sriyer@google.com>

### CODE REVIEW

**Comment by sriyerg** [comment]
Here's the advantage this solution brings in. In the example below, 2 tests are invoked in parallel by our batch regression runner `dvsim`. 

The first one compiles the `hello_world`:
```console
...
[47/49] Linking target sw/device/boot_rom/boot_rom_sim_dv.elf.
[48/49] Generating boot_rom_sim_dv with a custom command.
[49/49] Generating boot_rom_export_sim_dv with a custom command.
...
[12/14] Linking target sw/device/examples/hello_world/hello_world_sim_dv.elf.
[13/14] Generating hello_world_sim_dv with a custom command.
[14/14] Generating hello_world_export_sim_dv with a custom command.
```

Generating the boot_rom image required 49 steps (for compiling all the dependencies). Generating the hello_world image required 14 steps (since it shares several of the libraries with the boot_rom image). 

When running the `flash_test` that was invoked in parallel, it waited to acquire the lock. When the lock was released, the boot_rom required only 1 step and the flash_test, only 6 steps. This highlights the advantage of the speed up we get by building the SW in a common area rather than uniquely for each test. 
 
```console
...
[1/1] Generating boot_rom_export_sim_dv with a custom command.
...
[4/6] Linking target sw/device/tests/flash_ctrl/flash_test_sim_dv.elf.
[5/6] Generating flash_test_sim_dv with a custom command.
[6/6] Generating flash_test_export_sim_dv with a custom command.
```

**Comment by tjaychen** [comment]
this is great! i need to look at in more detail...but i did have one question... how would this work overall, if we ever have differently parameterized hardware and that somehow translates into different sw lib requirements?

I don't think we really have such a case right now..unless we have a regression that is meant to sweep through a bunch of xbar configs.

**Comment by sriyerg** [comment]
> this is great! i need to look at in more detail...but i did have one question... how would this work overall, if we ever have differently parameterized hardware and that somehow translates into different sw lib requirements?
> 
> I don't think we really have such a case right now..unless we have a regression that is meant to sweep through a bunch of xbar configs.

My hope is parameterizations are resolved at runtime as opposed to compile time. The same UART DIF library for example, will cater to all parametrized instances in the chip (so it would need to be compiled into an obj only once). Based on the instance, the DIF code will have safeguards in place to enable a feature only if it is available in that instance.

**Comment by rswarbrick** [comment]
Is the procmail dependency actually needed here? If you just need a lock file to serialize jobs, I would have thought you could do this equally well with [flock (1)](http://man7.org/linux/man-pages/man1/flock.1.html), which doesn't pull in a dependency.

**Comment by mcy** [comment]
I am mildly suspicious as to whether this speedup actually does what you want; all of the threads are going to contend for the single build task, and essentially wait for it to complete before starting.

Wouldn't it make more sense to pre-build *all* the software before you run your tests? Your tradeoff looks something like this:
- Pre-build everything. This means all tests pay for the linking cost of all other tests.
- Have each text build everything from scratch. This means that all tests pay for the same common `.c -> .o` compilation.

I would think that the linking cost is lower, and it avoids having to think about lock contention. I think you should benchmark both of these approaches before doing something complicated with locks.

File: hw/dv/data/sim.mk [hunk 1]
```diff
@@ -4,6 +4,9 @@
 
 .DEFAULT_GOAL := all
 
+LOCK_SW_BUILD   ?= lockfile -1 -l 3600 ${sw_build_dir}/lock
+UNLOCK_SW_BUILD ?= rm -rf ${sw_build_dir}/lock
```
Comment (gkelly) [inline hw/dv/data/sim.mk:8]: Why `-rf`? The created lockfile is a normal file isn't it?

**Comment by sriyerg** [comment]
> Is the procmail dependency actually needed here? If you just need a lock file to serialize jobs, I would have thought you could do this equally well with [flock (1)](http://man7.org/linux/man-pages/man1/flock.1.html), which doesn't pull in a dependency.

I had tried `flock` first and for whatever reason it hadn't work properly at that time (I may have skipped clearing the builds in between the trial runs). I have switched back to using `flock` - it does work well as intended.

File: hw/dv/data/sim.mk (hunk 1 above)
Comment (sriyerg) [reply hw/dv/data/sim.mk]: This part is now rewritten.

**Comment by sriyerg** [comment]
> I am mildly suspicious as to whether this speedup actually does what you want; all of the threads are going to contend for the single build task, and essentially wait for it to complete before starting.
> 
> Wouldn't it make more sense to pre-build _all_ the software before you run your tests? Your tradeoff looks something like this:
> 
> * Pre-build everything. This means all tests pay for the linking cost of all other tests.
> * Have each text build everything from scratch. This means that all tests pay for the same common `.c -> .o` compilation.
> 
> I would think that the linking cost is lower, and it avoids having to think about lock contention. I think you should benchmark both of these approaches before doing something complicated with locks.

The threads will indeed contend on the sw build step. The first thread that is granted will take the longest to compile all libraries. Subsequently, remaining threads wait for the previous ones to finish and pay only the linking cost of previous tests. This is still better than pre-build-all-tests approach where all tests wait for all builds to complete (with this approach, simulation starts immediately as each SW test is build). At worst, this has a very similar performance as the pre-build-all-tests approach. Plus, we are building only the tests we want to run, rather than building all of the available tests, which is not a scalable approach (example, 100s of SW tests getting unnecessarily built when we want to run only 1, or a smaller subset). 

I would like to avoid building all tests from scratch - that just feels a little wasteful to me.

File: apt-requirements.txt [hunk 2]
```diff
@@ -21,6 +21,7 @@ libusb-1.0-0
 make
 ninja-build
 pkgconf
+procmail
```
[the remark this answers is not in the document]
Comment (sriyerg) [reply apt-requirements.txt]: Removed.

**Comment by sriyerg** [comment]
> Other than my contention concerns, I think this is fine. I don't have a real hose in this race, since this this doesn't touch Meson directly.
> 
> If we do wind up finding that this isn't as good of an improvement as we had hoped, we can try to think of something cleaner.

Agree - we can always revisit if this later if needed.

**Pull request merged**